src/hotspot/cpu/x86/interp_masm_x86.cpp

```diff
@@ -973,9 +1004,57 @@ void InterpreterMacroAssembler::remove_activation(
 
     bind(no_reserved_zone_enabling);
   }
-  leave();                           // remove frame anchor
-  pop(ret_addr);                     // get return address
-  mov(rsp, rbx);                     // set sp to sender sp
+
+  // For asynchronous profiling to work correctly, we must remove the
+  // activation frame _before_ we test the method return safepoint poll.
+  // This is equivalent to how it is done for compiled frames.
+  // Removing an interpreter activation frame from a sampling perspective means
+  // updating the frame link. But since we are unwinding the current frame,
+  // we must save the current rbp in a temporary register, current_fp, for use
+  // as the last java fp should we decide to unwind.
+  // The asynchronous profiler will only see the updated rbp, either using the
+  // CPU context or by reading the saved_Java_fp() field as part of the ljf.
+  const Register current_fp = rscratch2;
+  const Register return_addr = rscratch2;
+  const Register continuation_return_pc = rscratch1;
+  const Register sender_sp = rscratch1;
+  movptr(return_addr, Address(rbp, wordSize)); // return address
+  // Load address of ContinuationEntry return pc
+  lea(continuation_return_pc, ExternalAddress(ContinuationEntry::return_pc_address()));
+  // Load the ContinuationEntry return pc
+  movptr(continuation_return_pc, Address(continuation_return_pc));
+  Label L_continuation, L_end;
+  cmpptr(continuation_return_pc, return_addr);
+  movptr(current_fp, rbp); // Save current fp in temporary register.
+  jcc(Assembler::equal, L_continuation);
+  movptr(rbp, Address(rbp, frame::link_offset)); // Update the frame link.
+  jmp(L_end);
+  bind(L_continuation);
+  lea(sender_sp, Address(rbp, frame::sender_sp_offset* wordSize));
+  movptr(rbp, Address(sender_sp, (int)(ContinuationEntry::size()))); // Update the frame link.
+  bind(L_end);
+
+  // The interpreter frame is now unwound from a sampling perspective,
+  // meaning it sees the sender frame as the current frame from this point onwards.
+
+  // The below poll is for the stack watermark barrier. It allows fixing up frames lazily,
+  // that would normally not be safe to use. Such bad returns into unsafe territory of
+  // the stack, will call InterpreterRuntime::at_unwind.
+  Label slow_path;
+  Label fast_path;
+  safepoint_poll(slow_path, rthread, current_fp, true /* at_return */, false /* in_nmethod */);
```

Comment:
A little heads-up: I am going to propose a little cleanup soon to drop `rthread` from x86 safepoint_pool (we can trust it is `r15_thread` always). That would probably yield a minor merge conflict here.

Reply:
FYI, it would be here: https://github.com/openjdk/jdk/pull/24323